Admission policies + some extra stats for cachebench

byrnedj opened the pull request:
This is the first part of what was formerly the background data movement + tier admission PR

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/intel/cachelib/10)
<!-- Reviewable:end -->

igchor committed:
Initial multi-tier support implementation

igchor committed:
Adjust and enable tests for ShmFileSegment

igchor committed:
Introduce FileShmSegment for file-backed shared memory

It's implementation is mostly based on PosixShmSegment.

Also, extend ShmManager and ShmSegmentOpts to support this new
segment type.

guptask committed:
Add support for shm opts serialization

After introducing file segment type, nameToKey_ does not provide
enough information to recover/remove segments on restart.

This commit fixes that by replacing nameToKey_ with nameToOpts_.

Previously, the Key from nameToKey_ map was only used in a single
DCHECK().

victoria-mcgrath committed:
Initial version of config API extension to support multiple memory tiers

* New class MemoryTierCacheConfig allows to configure a memory tier.
  Setting tier size and location of a file for file-backed memory are
  supported in this initial implementation;
* New member, vector of memory tiers, is added to class CacheAllocatorConfig.
* New test suite, chelib/allocator/tests/MemoryTiersTest.cpp,
  demonstrates the usage of and tests extended config API.

igchor committed:
Integrate Memory Tier config API with CacheAllocator.

igchor committed:
Run centos and debian workflows on push and PR

igchor committed:
Run tests on CI

igchor committed:
Add MemoryTierCacheConfig::fromShm()

to allow using new configureMemoryTiers() API with legacy behavior.

Move validation code for memory tiers to validate() method and convert
ratios to sizes lazily (on get)..

guptask committed:
Moved common segment code for posix and file shm segments into ShmCommon

igchor committed:
Fix test_shm_manager.cpp test

It wrongly assumed that the only possible segment type is
PosixSysV segment.

igchor committed:
Run long tests (navy/bench) every day on CI

victoria-mcgrath committed:
Enabled memory tier config API for cachebench.

victoria-mcgrath committed:
Enabled shared memory tier in cachebench.

Commit:
Adding example for multitiered cache

igchor committed:
Extend CompressedPtr to work with multiple tiers

Now it's size is 8 bytes intead of 4.

Original CompressedPtr stored only some offset with a memory Allocator.
For multi-tier implementation, this is not enough. We must also store
tierId and when uncompressing, select a proper allocator.

An alternative could be to just resign from CompressedPtr but they
are leveraged to allow the cache to be mapped to different addresses on shared memory.

Changing CompressedPtr impacted CacheItem size - it increased from 32 to 44 bytes.

igchor committed:
Run CI on prebuild docker image

igchor committed:
Run only centos build on CI

Commit:
Implemented async Item movement between tiers

igchor committed:
Enable workarounds in tests

igchor committed:
Add basic multi-tier test

igchor committed:
Set correct size for each memory tier

Commit:
Aadding new configs to hit_ratio/graph_cache_leader_fobj

Commit:
Fix eviction flow and removeCb calls

Without this fix removeCb called even in case when Item is moved between
tiers.

igchor committed:
Remove failing build-cachelib workflow (#42)

It fails because CentOS is EOL. We might want to consider
using CentOS Streams but for now, just remove it.

Right now, we rely on build-cachelib-centos workflow anyway.

victoria-mcgrath committed:
Disabled test suite allocator-test-AllocatorTypeTest (#41)

Disabled test suite allocator-test-AllocatorTypeTest to skip sporadically failing tests.

igchor committed:
Do not compensate for rounding error when calculating tier sizes (#43)

Compensation results in ratios being different than originially specified.

victoria-mcgrath committed:
Fixed total cache size in CacheMemoryStats (#38)

Return a sum of sizes of each tier instead of just 1st tier's size.

igchor committed:
Update docker file used in CI

Centos8 is EOL

igchor committed:
Disable failing clang-format-check

igchor committed:
Add one more navy test to BLACKLIST

Commit:
Fix issue with "Destorying an unresolved handle"

The issue happened when ReadHandleImpl ctor needs to destroy
waitContext_ because addWaitContextForMovingItem() returns false.
So before destroying waitContext_ we are calling discard method to
notify ~ItemWaitContext() that Item is ready.

igchor committed:
Add scripts for rebuilding/pushing docker images

Taken from: https://github.com/pmem/dev-utils-kit/commit/30794c3e1bbc9273e87da3e8f3ce7e5a2792b19e

igchor committed:
Extend CI to rebuild docker automatically

igchor committed:
Add extra param to build-package.sh

igchor committed:
Shorten critical section in findEviction

Remove the item from mmContainer and drop the lock before attempting
eviction.

Use moving bit for synchronization in findEviction

moving bit is used to give exclusive right to evict the item
to a particular thread.

Originially, there was an assumption that whoever marked the item
as moving will try to free it until he succeeds. Since we don't want
to do that in findEviction (potentially can take a long time) we need
to make sure that unmarking is safe.

This patch checks the flags after unmarking (atomically) and if ref is
zero it also recyles the item. This is needed as there might be some
concurrent thread releasing the item (and decrementing ref count). If
moving bit is set, that thread would not free the memory back to
allocator, resulting in memory leak on unmarkMoving().

mcengija committed:
Added required packages to install Intel ittapi

igchor committed:
Update build-cachelib-docker.yml

Do not use shallow clone to make sure Docker rebuild logic works correctly.

igchor committed:
Fix slab release code

Get tier id of item before calling any function on allocator
(which needs the tierID).

igchor committed:
critical section inside combined_lock

igchor committed:
Enable touchValue by default

igchor committed:
Issue75 rebased (#88)

* #75: Use actual tier sizes (rounded down to slab size and decreased by header size) when creating new memory pools

* Added getPoolSize method to calculate combined pool size for all tiers; added pool size validation to tests

* Explicitly specified type for totalCacheSize to avoid overflow

* Minor test change

* Reworked tests

* Minor change

* Deleted redundant tests

* Deleted unused constant

* First set of changes to cache configuration API to enable multi-tier caches (#138)

Summary:
These changes introduce per-tier cache configuration required to implement features discussed here: https://github.com/facebook/CacheLib/discussions/102. These specific changes enable single DRAM tier configs only which are compatible with the current version of cachelib. Configuration API will be expanded as multi-tier changes in other parts of the library are introduced.

Pull Request resolved: https://github.com/facebook/CacheLib/pull/138

Reviewed By: therealgymmy

Differential Revision: D36189766

Pulled By: jiayuebao

fbshipit-source-id: 947aa0cd800ea6accffc1b7b6b0c9693aa7fc0a5

Co-authored-by: Victoria McGrath <victoria.mcgrath@intel.com>

igchor committed:
Add memory usage statistics for slabs and allocation classes

igchor committed:
Add option to print memory stats in bytes only

guptask committed:
added per tier pool class rolling average latency

byrnedj committed:
MM2Q promotion iterators (#1)

Hot queue iterator for 2Q. Will start at Hot queue and move to Warm queue if hot queue is exhausted. Useful for promotion semantics if using 2Q replacement. rebased on to develop and added some tests.

byrnedj committed:
tier admission

item movement test

updated class in test

use transparent sync for item movement

remove extra whitespace

updates per comments

updates per comments (still outstanding on the todos)

byrnedj committed:
use transparent sync for item movement

Commit:
Implement NUMA binding support for SysVShmSegment

byrnedj committed:
remove extra whitespace

Commit:
Extend cachebench to bind memory tiers to NUMA nodes

Commit:
Adding AllocatorMemoryTiersTest for NUMA bindings

Commit:
Implement NUMA binding support for PosixShmSegment

Commit:
Fix deprecation warning in multitier example

vinser52 reviewed (changes requested):
> Reviewed 1 of 24 files at r1.
> *[Reviewable](https://reviewable.io/reviews/intel/cachelib/10)* status: 1 of 24 files reviewed, 13 unresolved discussions (waiting on @byrnedj and @vinser52)

___
*[`cachelib/allocator/Cache.h` line 100 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyY0tVCCqo9BjHReFJ:-NAyY0tVCCqo9BjHReFK:b-s3w12d) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/Cache.h#L100)):*
> ```objc
>   //
>   // @param poolId    The pool id to query
>   // @param tierId    The tier of the pool id
> ```

Should we extend the comment for the method above (that takes only `poolId`) that it returns a reference to the top-most tier memory pool? 
___
*[`cachelib/allocator/CacheAllocator.h` line 39 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyYOjr2MdkaM-x0NHq:-NAyYOjr2MdkaM-x0NHr:b-41njf1) ([raw file](https://github.com/intel/cachelib/blob/9f5dc7c00e08f219da98c45345be14722a7331a2/cachelib/allocator/CacheAllocator.h#L39)):*
> ```objc
> #include <folly/Range.h>
> #pragma GCC diagnostic pop
> ```

I would keep empty line to avoid merge conflicts in the future.
___
*[`cachelib/allocator/CacheAllocator-inl.h` line 376 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAybbu16YtBsTJvXw0s:-NAybbu16YtBsTJvXw0t:b-y05mmh) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/CacheAllocator-inl.h#L376)):*
> ```objc
>                                              uint32_t creationTime,
>                                              uint32_t expiryTime,
>                                              bool fromEvictorThread) {
> ```

Where the `fromEvictorThread` variable is used?
___
*[`cachelib/allocator/CacheAllocator-inl.h` line 393 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyaVc13WeKXOfH7v5G:-NAyaVc13WeKXOfH7v5H:b9yonq0) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/CacheAllocator-inl.h#L393)):*
> ```objc
>   void *memory = nullptr;
> 
>   if (tid == 0 && config_.acTopTierEvictionWatermark > 0.0
> ```

Since we are using some heuristics here, I think a comment is needed which describes the logic.

Why do we try to findEviction first? Even if we have less free space than threshhold I think it is better to try to allocate in this tier rather than having an eviction on the hot path. Or I do not understand the logic. So a comment is really desirable. 
___
*[`cachelib/allocator/CacheAllocator-inl.h` line 476 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAycGH33h0V5NgoeA07:-NAycGH33h0V5NgoeA08:b-fhommq) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/CacheAllocator-inl.h#L476)):*
> ```objc
> 
>   if (freePercentage <= config_.minAcAllocationWatermark)
>     return defaultTargetTier + 1;
> ```

But it might have even less free space... I thought `freePercentage` is a parameter that BG evictor monitor to decided whether it should evict something from this tier to reach target `freePercentage`, isn't it?
___
*[`cachelib/allocator/CacheAllocator-inl.h` line 478 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAycjhh3MYSXezmNFJY:-NAycjhh3MYSXezmNFJZ:b-zgqens) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/CacheAllocator-inl.h#L478)):*
> ```objc
>     return defaultTargetTier + 1;
> 
>   // TODO: we can even think about creating different allocation classes for PMEM
> ```

We should avoid `PMEM` term.
___
*[`cachelib/allocator/CacheAllocator-inl.h` line 500 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyd2RM-vqaLLkdsCHy:-NAyd2RNAQ0J1Ghm1-HC:b-ks2w6s) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/CacheAllocator-inl.h#L500)):*
> ```objc
> 
>   // TODO: only works for 2 tiers
>   return (folly::Random::rand32() % 100) < config_.defaultTierChancePercentage ? defaultTargetTier : defaultTargetTier + 1;
> ```

Not sure about folly implementation but frequently rand function has an internal lock and having it on a hot path might brake concurrency
___
*[`cachelib/allocator/CacheAllocator-inl.h` line 1683 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAydWGC-vozpfKfgTK4:-NAydWGC-vozpfKfgTK5:bsm4c6w) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/CacheAllocator-inl.h#L1683)):*
> ```objc
>     TierId sourceTierId, TierId targetTierId, PoolId pid, Item& item)
> {
>   if (config_.disableEvictionToMemory)
> ```

it could be `return !(config_.disableEvictionToMemory)` to avoid branch.
___
*[`cachelib/allocator/CacheAllocatorConfig.h` line 594 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyds2I5Dm9wV9GReMb:-NAyds2I5Dm9wV9GReMc:b-t1jrin) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/CacheAllocatorConfig.h#L594)):*
> ```objc
>   double acTopTierEvictionWatermark{0.0}; // TODO: make it per TIER?
>   uint64_t sizeThresholdPolicy{0};   
>   double defaultTierChancePercentage{50.0};
> ```

I think new config option should somehow described in the comments.
___
*[`cachelib/allocator/MemoryTierCacheConfig.h` line 72 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAye4mmD4yawkN90N6S:-NAye4mn9cWG8xCV_-L0:b-6k9cji) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/MemoryTierCacheConfig.h#L72)):*
> ```objc
>   }
> 
>     // TODO: move it to MMContainer config
> ```

Looks like we need to address this TODO comment. Not sure why it should be per tier.
___
*[`cachelib/allocator/memory/MemoryAllocator.h` line 652 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyfgNgCVMlE0Nr-Kwo:-NAyfgNgCVMlE0Nr-Kwp:b-gsf26j) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/memory/MemoryAllocator.h#L652)):*
> ```objc
>   }
> 
>  // TODO:
> ```

what is it?
___
*[`cachelib/allocator/memory/MemoryAllocatorStats.h` line 58 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyfqAJ2cQ5IqXYfHst:-NAyfqAJ2cQ5IqXYfHsu:b-4zgfy6) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/memory/MemoryAllocatorStats.h#L58)):*
> ```objc
>   }
> 
>   constexpr size_t getTotalMemory() const noexcept {
> ```

should it be renamed to `getTotalAllocMemory`?
___
*[`cachelib/allocator/memory/MemoryPool.h` line 311 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyfxbQEgayMCVuHQ-C:-NAyfxbQEgayMCVuHQ-D:b-9q20i4) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/memory/MemoryPool.h#L311)):*
> ```objc
>   void setNumSlabsAdvised(uint64_t value) { curSlabsAdvised_ = value; }
> 
>  // TODO:
> ```

Again, we broken incapsulation. Not sure that we have to merge some hacks to develop branch.


<!-- Sent from Reviewable.io -->

vinser52 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/intel/cachelib/10)* status: 1 of 24 files reviewed, 15 unresolved discussions (waiting on @byrnedj and @vinser52)

___
*[`cachelib/allocator/tests/AllocatorMemoryTiersTest.cpp.orig` line 1 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAygF5pD4j1s7fgtxeU:-NAygF5pD4j1s7fgtxeV:bs4n0aa) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/tests/AllocatorMemoryTiersTest.cpp.orig#L1)):*
> ```
> /*
> ```

why are these files part of PR?
___
*[`cachelib/allocator/tests/AllocatorMemoryTiersTest.h.orig` line 1 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAygMRz2Ha56lXOFStL:-NAygMRz2Ha56lXOFStM:bs4n0aa) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/tests/AllocatorMemoryTiersTest.h.orig#L1)):*
> ```
> /*
> ```

why are these files part of PR?


<!-- Sent from Reviewable.io -->

Commit:
Merge pull request #15 from vinser52/fix_deprecation_warn

Fix deprecation warn

Commit:
Update CI to use intel/CacheLib repo (#17)

Commit:
Clean-up test's blacklist (#18)

Commit:
Merge pull request #13 from vinser52/numa_binding

Initial support of NUMA bindings

Commit:
Fix ReaperSkippingSlabTraversalWhileSlabReleasing test

The issue was caused by incorrect behaviour of the
CacheAllocator<CacheTrait>::tryEvictToNextMemoryTier method in case the
evicted item is expired. We cannot simply return a handle to it, but we need
to remove it from the access container and MM container.

Commit:
Merge pull request #19 from vinser52/fix_allocator_test

Fix ReaperSkippingSlabTraversalWhileSlabReleasing test

byrnedj commented:
I rebased the PR, the outstanding issues from the review to address are:

[cachelib/allocator/MemoryTierCacheConfig.h line 72 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAye4mmD4yawkN90N6S:-NAye4mn9cWG8xCV_-L0:b-6k9cji) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/MemoryTierCacheConfig.h#L72)):

  }

    // TODO: move it to MMContainer config
Looks like we need to address this TODO comment. Not sure why it should be per tier.

[cachelib/allocator/memory/MemoryAllocator.h line 652 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyfgNgCVMlE0Nr-Kwo:-NAyfgNgCVMlE0Nr-Kwp:b-gsf26j) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/memory/MemoryAllocator.h#L652)):

  }

 // TODO:
what is it?

[cachelib/allocator/memory/MemoryAllocatorStats.h line 58 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyfqAJ2cQ5IqXYfHst:-NAyfqAJ2cQ5IqXYfHsu:b-4zgfy6) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/memory/MemoryAllocatorStats.h#L58)):

  }

  constexpr size_t getTotalMemory() const noexcept {
should it be renamed to getTotalAllocMemory?

[cachelib/allocator/memory/MemoryPool.h line 311 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyfxbQEgayMCVuHQ-C:-NAyfxbQEgayMCVuHQ-D:b-9q20i4) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/memory/MemoryPool.h#L311)):

  void setNumSlabsAdvised(uint64_t value) { curSlabsAdvised_ = value; }

 // TODO:
Again, we broken incapsulation. Not sure that we have to merge some hacks to develop branch.

igchor reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/intel/cachelib/10)* status: 0 of 24 files reviewed, 16 unresolved discussions (waiting on @byrnedj, @igchor, and @vinser52)

___
*[`cachelib/allocator/CacheAllocator-inl.h` line 393 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyaVc13WeKXOfH7v5G:-NCRZQD2-H4kdv6X-bT-:b-8c4eem) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/CacheAllocator-inl.h#L393)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

Since we are using some heuristics here, I think a comment is needed which describes the logic.

Why do we try to findEviction first? Even if we have less free space than threshhold I think it is better to try to allocate in this tier rather than having an eviction on the hot path. Or I do not understand the logic. So a comment is really desirable. 
</blockquote></details>

Yes, by default we should allocate in this tier (hence acTopTierEvictionWatermark is set to 0 in config). However, this condition allows you to change this behavior if you want (which can allow us to test performance of that).
___
*[`cachelib/allocator/CacheAllocator-inl.h` line 476 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAycGH33h0V5NgoeA07:-NCR_-HlDkXb3V9rYQXw:b-7fch08) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/CacheAllocator-inl.h#L476)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

But it might have even less free space... I thought `freePercentage` is a parameter that BG evictor monitor to decided whether it should evict something from this tier to reach target `freePercentage`, isn't it?
</blockquote></details>

Yes, BG monitors it, but we can still use it for comparison here. What do yoy mean by "it might have even less free space"?
___
*[`cachelib/allocator/CacheAllocator-inl.h` line 2508 at r2](https://reviewable.io/reviews/intel/cachelib/10#-NCR_OCe5XSabzOogFk0:-NCR_OCe5XSabzOogFk1:b-6sx5o0) ([raw file](https://github.com/intel/cachelib/blob/3181c63d6b8366565b77233310ab0d49ebc59ac3/cachelib/allocator/CacheAllocator-inl.h#L2508)):*
> ```objc
>             : 0);
>     for (TierId tid = 0; tid < getNumTiers(); tid++) {
>       if constexpr (std::is_same_v<MMConfig, MMLru::Config> || std::is_same_v<MMConfig, MM2Q::Config>) {
> ```

This should probably be fixed (this if was just a temp workaround).
The best way would probably be to move those 2 params to specific MMContainer config. Do we have support for setting MMCOntainer configs from JSON currently? And will it work for multi tier?
___
*[`cachelib/allocator/memory/MemoryAllocator.h` line 652 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyfgNgCVMlE0Nr-Kwo:-NCRa0xtEFyAIF_7ttWv:b-cd7x7m) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/memory/MemoryAllocator.h#L652)):*
<details><summary><i>Previously, byrnedj (Daniel Byrne) wrote…</i></summary><blockquote>

  }

 // TODO:
what is it?

</blockquote></details>

`private` should be uncommented - if anything breaks, you need to add some extra function to the MemoryAllcoator
___
*[`cachelib/allocator/memory/MemoryPool.h` line 311 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyfxbQEgayMCVuHQ-C:-NCRa_jD9TovkSn05zP0:bjtne0b) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/memory/MemoryPool.h#L311)):*
<details><summary><i>Previously, byrnedj (Daniel Byrne) wrote…</i></summary><blockquote>

  void setNumSlabsAdvised(uint64_t value) { curSlabsAdvised_ = value; }

 // TODO:
Again, we broken incapsulation. Not sure that we have to merge some hacks to develop branch.

</blockquote></details>

Agree, `private` should be uncommented - if anything breaks, you need to add some extra function to the MemoryPool


<!-- Sent from Reviewable.io -->

byrnedj committed:
updated per review

vinser52 reviewed (changes requested):
> Reviewed 2 of 23 files at r2.
> *[Reviewable](https://reviewable.io/reviews/intel/cachelib/10)* status: 2 of 24 files reviewed, 16 unresolved discussions (waiting on @byrnedj, @igchor, and @vinser52)

___
*[`cachelib/allocator/CacheAllocator.h` line 1291 at r3](https://reviewable.io/reviews/intel/cachelib/10#-NCX-SqP98XHgS2DIvaG:-NCX-SqP98XHgS2DIvaH:b-z0u3eu) ([raw file](https://github.com/intel/cachelib/blob/f08e43480b837b88f141695923bf85735acf1719/cachelib/allocator/CacheAllocator.h#L1291)):*
> ```objc
>   CacheAllocator(InitMemType types, Config config);
>   
>   TierId getTargetTierForItem(PoolId pid, typename Item::Key key,
> ```

I think we need a comment as for other methods in this class to desccribe the logic of this method.
___
*[`cachelib/allocator/CacheAllocator.h` line 1686 at r3](https://reviewable.io/reviews/intel/cachelib/10#-NCX-dm6B2lxOYBJj9dU:-NCX-dm6B2lxOYBJj9dV:b-12c1lm) ([raw file](https://github.com/intel/cachelib/blob/f08e43480b837b88f141695923bf85735acf1719/cachelib/allocator/CacheAllocator.h#L1686)):*
> ```objc
>   WriteHandle tryEvictToNextMemoryTier(Item& item);
> 
>   bool shouldEvictToNextMemoryTier(TierId sourceTierId,
> ```

The same as previous, the comment is needed.
___
*[`cachelib/allocator/CacheAllocator-inl.h` line 393 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyaVc13WeKXOfH7v5G:-NCWzwId8JIXmPXzVsou:b94d3kb) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/CacheAllocator-inl.h#L393)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

Yes, by default we should allocate in this tier (hence acTopTierEvictionWatermark is set to 0 in config). However, this condition allows you to change this behavior if you want (which can allow us to test performance of that).
</blockquote></details>

Ah, got it. Than some comment is nice to have to describe this logic.

BTW, do we have cases when this option improves performance somehow?
___
*[`cachelib/allocator/CacheAllocator-inl.h` line 476 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAycGH33h0V5NgoeA07:-NCX02ht3WjX0TeZ85U2:bm9ksly) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/CacheAllocator-inl.h#L476)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

Yes, BG monitors it, but we can still use it for comparison here. What do yoy mean by "it might have even less free space"?
</blockquote></details>

As I understood, if default tier has no enough space than we return next tier. But what if next tier has even less free space?
Another question: will such logic works for cases with more than two tiers? I understand that today we are limiting number of tiers to 2. But what if the constant will be changed?
___
*[`cachelib/allocator/CacheAllocator-inl.h` line 528 at r3](https://reviewable.io/reviews/intel/cachelib/10#-NCX16XWEo9OVeg7yOl0:-NCX16XWEo9OVeg7yOl1:b-38ume7) ([raw file](https://github.com/intel/cachelib/blob/f08e43480b837b88f141695923bf85735acf1719/cachelib/allocator/CacheAllocator-inl.h#L528)):*
> ```objc
>                                              uint32_t expiryTime) {
>   auto tid = getTargetTierForItem(pid, key, size, creationTime, expiryTime);
>   return allocateInternalTier(tid, pid, key, size, creationTime, expiryTime);
> ```

Is it safe to remove the loop? I can imagine situation when getTargetTierForItem return tier 0. After that concurrent threads fills the tier 0 and make it full. After that our thread will be unable to allocate in tier 0 and return empty handle. I understand it is crazy situation, but what if? 
___
*[`cachelib/allocator/CacheAllocator-inl.h` line 2566 at r3](https://reviewable.io/reviews/intel/cachelib/10#-NCX2Fjh4yM6Igh-UNM6:-NCX2Fjh4yM6Igh-UNM7:b-5qbztg) ([raw file](https://github.com/intel/cachelib/blob/f08e43480b837b88f141695923bf85735acf1719/cachelib/allocator/CacheAllocator-inl.h#L2566)):*
> ```objc
>   // TODO - get rid of the duplication - right now, each tier
>   // holds pool objects with mostly the same info
>   return filterCompactCachePools(allocator_[currentTier()]->getPoolIds());
> ```

@igchor should it be really currentTier() instead of always 0? As I remember you added currentTier() to be used for interaction with BG workers. In this case set of pools is the same for each tier (we have duplication) and we can always work with tier 0, isn’t it?
___
*[`cachelib/allocator/MemoryTierCacheConfig.h` line 72 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAye4mmD4yawkN90N6S:-NCX2x3m5lMRNa9vJoIX:b5mukox) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/MemoryTierCacheConfig.h#L72)):*
<details><summary><i>Previously, byrnedj (Daniel Byrne) wrote…</i></summary><blockquote>

  }

    // TODO: move it to MMContainer config
Looks like we need to address this TODO comment. Not sure why it should be per tier.

</blockquote></details>

Not sure that I understand, but we have MMContainers per tier but config for MMcontainers can be the same for each tier. Is my assumption correct?


<!-- Sent from Reviewable.io -->

byrnedj committed:
added admission policy doc

igchor reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/intel/cachelib/10)* status: 2 of 25 files reviewed, 16 unresolved discussions (waiting on @byrnedj, @igchor, and @vinser52)

___
*[`cachelib/allocator/CacheAllocator-inl.h` line 476 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAycGH33h0V5NgoeA07:-NCb5fjy2dcidkCop776:b-tdr7w1) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/CacheAllocator-inl.h#L476)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

As I understood, if default tier has no enough space than we return next tier. But what if next tier has even less free space?
Another question: will such logic works for cases with more than two tiers? I understand that today we are limiting number of tiers to 2. But what if the constant will be changed?
</blockquote></details>

1. Yes, it might happen that next tier has even less space but at least this option gives you choice - right now we always allocate in topmost (and with this patch, default setting is the same - i.e. maxAcAllocationWatermark is 0). In general, we can try to apply some more dynamic threshold, use hill climbing approach, etc
2. Right, currently this assumes there are only 2 tiers (some other config options do that as well). I think that if we would like to have this generic then we would have to introduce such allocationWatermarks per tier (or per each boundry between tiers)
___
*[`cachelib/allocator/CacheAllocator-inl.h` line 2508 at r2](https://reviewable.io/reviews/intel/cachelib/10#-NCR_OCe5XSabzOogFk0:-NCb6dwi3e9TLzwWaKiS:bxnwjgl) ([raw file](https://github.com/intel/cachelib/blob/3181c63d6b8366565b77233310ab0d49ebc59ac3/cachelib/allocator/CacheAllocator-inl.h#L2508)):*
<details><summary><i>Previously, byrnedj (Daniel Byrne) wrote…</i></summary><blockquote>

For this it is not trivial to add per tier mmContainer configs. Partly because in cachebench/Cache-inl.h, we have the following initialization:
```    
    const size_t numBytes = cache_->getCacheMemoryStats().cacheSize;
    for (uint64_t i = 0; i < config_.numPools; ++i) {
      const double& ratio = config_.poolSizes[i];
      const size_t poolSize = static_cast<size_t>(numBytes * ratio);
      typename Allocator::MMConfig mmConfig =
          makeMMConfig<typename Allocator::MMConfig>(config_);
      const PoolId pid = cache_->addPool(
          folly::sformat("pool_{}", i), poolSize, {} /* allocSizes */, mmConfig,
          nullptr /* rebalanceStrategy */, nullptr /* resizeStrategy */,
          true /* ensureSufficientMem */);
      pools_.push_back(pid);
    }
```

The call to addPool initializes a pool that is split among tiers, so mmConfig is done per pool rather than per tier. For now I would suggest that we either:
1. Accept the current implementation or
2. Drop it from this PR and create a new PR that fully supports per-tier mmContainer creation (allowing us to have different allocators per tier too). 

I have addressed the rest of the comments in the latest commit. 

</blockquote></details>

I think we could drop the change, it might be worked on separately I belive.
___
*[`cachelib/allocator/CacheAllocator-inl.h` line 2566 at r3](https://reviewable.io/reviews/intel/cachelib/10#-NCX2Fjh4yM6Igh-UNM6:-NCb6r0yDgUxK2XJQ7MS:b-33iuvm) ([raw file](https://github.com/intel/cachelib/blob/f08e43480b837b88f141695923bf85735acf1719/cachelib/allocator/CacheAllocator-inl.h#L2566)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

@igchor should it be really currentTier() instead of always 0? As I remember you added currentTier() to be used for interaction with BG workers. In this case set of pools is the same for each tier (we have duplication) and we can always work with tier 0, isn’t it?
</blockquote></details>

Right, in this particular place, `0` is OK.
___
*[`cachelib/allocator/MemoryTierCacheConfig.h` line 72 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAye4mmD4yawkN90N6S:-NCb6xtz6KBwjJlIWG3K:bisdh64) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/MemoryTierCacheConfig.h#L72)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

Not sure that I understand, but we have MMContainers per tier but config for MMcontainers can be the same for each tier. Is my assumption correct?
</blockquote></details>

I belive we would prefer to have config for each MMContainer (for each tier separately). This is because we only evict from DRAM but we evict AND promote from the second tier (which means we might want to customize insertion point for second tier, but not neceserilay for DRAM),


<!-- Sent from Reviewable.io -->

igchor reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/intel/cachelib/10)* status: 2 of 25 files reviewed, 16 unresolved discussions (waiting on @byrnedj, @igchor, and @vinser52)

___
*[`cachelib/allocator/CacheAllocator-inl.h` line 393 at r1](https://reviewable.io/reviews/intel/cachelib/10#-NAyaVc13WeKXOfH7v5G:-NCb7MSHCG0Vp5qw5knw:bbkq36x) ([raw file](https://github.com/intel/cachelib/blob/138f33ad6fa700735076282a4a1e7e015be91eee/cachelib/allocator/CacheAllocator-inl.h#L393)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

Ah, got it. Than some comment is nice to have to describe this logic.

BTW, do we have cases when this option improves performance somehow?
</blockquote></details>

That's why I added this - to be able to benchmark to it :) 


<!-- Sent from Reviewable.io -->
